### PR TITLE
OpenVINO IR module

### PR DESCRIPTION
**Pull request by cih9088**
### Related OTX [PR](https://github.com/openvinotoolkit/training_extensions/pull/1399)
- OpenVINO IR to pytorch graph conversion
- OpenModelZoo downloader

**Need to add test code**

### CODE REVIEW

File: mpa/modules/ov/models/mmcls/heads/cls_head.py
Comment (harimkang) [inline mpa/modules/ov/models/mmcls/heads/cls_head.py:10]: This conflicts with the Head of the existing mmcls. we need to change the name
Or if there is no problem, we should `force=True`.

File: mpa/modules/ov/models/mmdet/necks/mmov_ssd_neck.py
Comment (harimkang) [inline mpa/modules/ov/models/mmdet/necks/mmov_ssd_neck.py]: The `mmdet` currently used in otx cannot be imported. If this is supported before mmpublic-det integration, it seems that this should be corrected.

File: mpa/modules/ov/models/mmdet/roi_heads/roi_extractors/single_level_roi_extractor.py
Comment (harimkang) [inline mpa/modules/ov/models/mmdet/roi_heads/roi_extractors/single_level_roi_extractor.py:16]: This class also conflicts with Extractor registered in the existing registry.

**Review by harimkang — COMMENTED** [review]
Based on [this PR](https://github.com/openvinotoolkit/training_extensions/pull/1399), I tried running sample with classification after reinstalling mpa as a branch of the current PR.
(otx/algorithms/init_venv.sh + mpa reinstalling as this PR)

I left a comment first on only the things needed to run the sample from current feature/otx

File: mpa/modules/ov/models/mmcls/heads/cls_head.py
Comment (cih9088) [reply mpa/modules/ov/models/mmcls/heads/cls_head.py:10]: Since this class merely adds negligible features, I believe `force=True` would be enough.
If you insisted to rename it, I would. Let me know.

File: mpa/modules/ov/models/mmdet/roi_heads/roi_extractors/single_level_roi_extractor.py
Comment (cih9088) [reply mpa/modules/ov/models/mmdet/roi_heads/roi_extractors/single_level_roi_extractor.py:16]: Since this class merely adds negligible features, I believe `force=True` would be enough.
If you insisted to rename it, I would. Let me know.

File: mpa/modules/ov/models/mmcls/heads/cls_head.py
Comment (harimkang) [reply mpa/modules/ov/models/mmcls/heads/cls_head.py:10]: I also think it's best to do `force=True` if this class doesn't have a big impact on the existing ones.

File: mpa/modules/ov/models/mmdet/necks/mmov_ssd_neck.py
Comment (cih9088) [reply mpa/modules/ov/models/mmdet/necks/mmov_ssd_neck.py]: Indeed, I assumed public `mmdet`. I would copy `SSDNeck` from public `mmdet` for the time being.

File: mpa/modules/ov/models/mmcls/backbones/mmov_backbone.py
Comment (harimkang) [inline mpa/modules/ov/models/mmcls/backbones/mmov_backbone.py:14]: The `MMOVBackbone` class of the same name is implemented for each task. Could it be possible to unify this? 
If merging is not possible, consider renaming the individual classes..
Also, how about registering this as mmcv's backbone?
you can refer this [code](https://github.com/openvinotoolkit/training_extensions/blob/b19ab2f359ce10885d0f602f37890a70e321945e/otx/algorithms/common/adapters/mmcv/models/builder.py#L20),
maybe.. we can use `BACKBONES = Registry("models", parent=MMCV_MODELS, scope="omz")` and then we can use `omz.MMOVBackbone` in model configuration.

Comment (cih9088) [reply mpa/modules/ov/models/mmcls/backbones/mmov_backbone.py:14]: First of all, we can't unify backbones since each library requires different outputs of backbone.
Could you explain why we need to name them differently?

I haven't seen, `omz.MMOVBackbone`, this kind of use case in OTX. Could you point it out?
What do you mean by **use omz.MMOVBackbone in model configuration.**?

Comment (harimkang) [reply mpa/modules/ov/models/mmcls/backbones/mmov_backbone.py:14]: Renaming was for my suggestion about registering in one registry of mmcv.

Recently `feature/otx`, It has been modified to register and find backbones in the registry in the form of `otx.OTXMobileNetV3`, `mmdet.ResNet` or `torchvision.resnet`. It was just my suggestion for public backbone support.

However, leave this as it is in this PR for now, and see if I can unify the format. (according to the style of otx)

**Review by goodsong81 — COMMENTED** [review]
@harimkang's comments are all for ease of otx find & build.
If it does not harm the functionality, let's align to some extent.
Of course, some part could be done in caller part. :)
In addition, @cih9088, we should not assume that "force=True" is forced in mmcv installation as we will remove init_venv.sh eventually.

**Comment by cih9088** [comment]
@harimkang 
After looking into [your PR](https://github.com/openvinotoolkit/training_extensions/pull/1391), I get the idea.
I have no objection to it but I believe registering `MMOVBackbone` to `mmcv`'s backbone should be implemented in OTX not MPA even though renaming should be done in MPA.

I have rebased onto current branch and fixed  bugs in https://github.com/openvinotoolkit/model_preparation_algorithm/pull/95/commits/249baf0852c7c94e9f93632528f44d779c08b792

File: mpa/modules/ov/models/__init__.py
Comment (harimkang) [inline mpa/modules/ov/models/__init__.py:1]: Why are we trying to import (mmcls/mmdet/mmseg) here?
(Currently, many warning logs appear when importing non-public (otx side)`mmseg`. I don't know exactly why... However, I mentioned it because this part seems to cause more logs.)

Comment (cih9088) [reply mpa/modules/ov/models/__init__.py:1]: It was to register each `MMOV` model to a library specific registry.
Could you share warning logs?

Comment (harimkang) [reply mpa/modules/ov/models/__init__.py:1]: Sorry for the confusion. I'm not sure if here is wrong. It seems to be importing `nncf.torch` from somewhere.. (Even though nncf isn't working, it looks like it's importing somewhere.)

and, Is this import work to check if import is possible in advance?
Shouldn't this be inside the folders in `mmcls`, `mmdet`, `mmseg` respectively?
For example.. I don't know if it's appropriate to import `mmdet` and `mmseg` once to use classification (This is probably enough for `mmcls`.).

**Review by harimkang — APPROVED** [review]
For me, Sample with this PR is working with OMZ model configuration for classification.
And when I checked the otx cli test locally using the corresponding MPA branch, there seemed to be no problem.

After the PR is merged, I plan to create a PR that allows me to configure and run the OMZ model through `otx find & build`, and I plan to add a test to check if it works by OMZ backbones.

@cih9088 This PR only supports classification first, right?

File: mpa/modules/ov/models/__init__.py
Comment (cih9088) [reply mpa/modules/ov/models/__init__.py:1]: I don't mind moving this test import to anywhere else. Your call.

**Comment by cih9088** [comment]
> For me, Sample with this PR is working with OMZ model configuration for classification. And when I checked the otx cli test locally using the corresponding MPA branch, there seemed to be no problem.
> 
> After the PR is merged, I plan to create a PR that allows me to configure and run the OMZ model through `otx find & build`, and I plan to add a test to check if it works by OMZ backbones.
> 
> @cih9088 This PR only supports classification first, right?

Even though some detection models and segmentation models are supported as well but, yes, OMZ download and convert features are only available for classification models for the time being.

**Comment by harimkang** [comment]
@cih9088 Hi. I have one question. Is there a need for different data pipelines for each backbone?

**Comment by cih9088** [comment]
> @cih9088 Hi. I have one question. Is there a need for different data pipelines for each backbone?

Which pipeline? I don't get it. Could you point it out in code and explain it a bit more?

**Comment by harimkang** [comment]
> Which pipeline? I don't get it. Could you point it out in code and explain it a bit more?

For example, when trying to use alexnet, I met the error below
image
and it works only when I change the value for resize to 227.
This size must be matched for each backbone to operate..? (It is correct that it is fixed, and this should be modified by the user now, right?)

**Comment by cih9088** [comment]
I see your problem. Here is a thing.
Some models have hard-coded values related to their input shape so that arbitrary input shape leads a wrong behaviour.
One could change model input to dynamic shapes with provided [OpenVINO API](https://docs.openvino.ai/latest/openvino_docs_OV_UG_DynamicShapes.html), but it has its own drawback.
`dynamic shape` API generates lots of new nodes assigned random name on them.
The number of nodes generated is not a problem. What matters is a random name of each generated nodes.
Because of the random names, a graph constructed each time one initialized `MMOV` series model is not identical so that we can not load `state_dict`, which means, after training, we can not load the model.

**Comment by goodsong81** [comment]
In my opinion, the input resolution should be model or use-case specific so that backbones should work in any resolution. I suppose backbone weights in *.bin files should be free from input resolution logically. On the other hand, value normalization is specific to the backbone, and IRs have their own input normalization layers in general. So we can omit input norm in case of OV backbone.

**Comment by goodsong81** [comment]
> I see your problem. Here is a thing. Some models have hard-coded values related to their input shape so that changing input shape resulting a wrong behaviour. One could change model input to dynamic shapes with provided [OpenVINO API](https://docs.openvino.ai/latest/openvino_docs_OV_UG_DynamicShapes.html), but it has its own drawback. `dynamic shape` API generates lots of new nodes assigned random name on them. The number of nodes generated is not a problem. What matters is a random name of each generated nodes. Because of the random names, a graph constructed each time one initialized `MMOV` series model is not identical so that we can not load `state_dict`, which means, after training, we can not load the model.

IR model needs reshape()-like API support but torch model does not.
Can we create torch model from IR in such way where resolution is adaptive to input?

**Comment by cih9088** [comment]
> > I see your problem. Here is a thing. Some models have hard-coded values related to their input shape so that changing input shape resulting a wrong behaviour. One could change model input to dynamic shapes with provided [OpenVINO API](https://docs.openvino.ai/latest/openvino_docs_OV_UG_DynamicShapes.html), but it has its own drawback. `dynamic shape` API generates lots of new nodes assigned random name on them. The number of nodes generated is not a problem. What matters is a random name of each generated nodes. Because of the random names, a graph constructed each time one initialized `MMOV` series model is not identical so that we can not load `state_dict`, which means, after training, we can not load the model.
> 
> IR model needs reshape()-like API support but torch model does not. Can we create torch model from IR in such way where resolution is adaptive to input?

Let me clarify it a bit further.
IR models from OMZ have a static input shape.
In some of models, the input shape is hard-coded in IR, not depends on input shape so that arbitrary input shape to model would cause a hard to debug error.

One can change this static shape to dynamic shape using [OpenVINO API](https://docs.openvino.ai/latest/openvino_docs_OV_UG_DynamicShapes.html) before converting to pytorch model. I believe the API would change all the hard-coded input shape thing.
However, the problem is that the API generates random named nodes.
This causes a graph mismatch every time you initialize the model, which means we can not load a trained model because a graph of trained model and a graph of newly initialize model are not identical.

Actually, the randomness is not really random. I think there is some sort of session thing in ngraph because each time when one applies `dynamic shape` API after completely quitting the process, generated random names are the same.
The problem arises when one uses `dynamic shape` API multiple times in one process.

**Comment by goodsong81** [comment]
I'm not talking about IR model reshape, but torch model.
(x) reshape IR to (w,h) -> convert to torch model
(o) convert to torch model -> infer on (w,h) input

I understand that IR models have static shape. But torch CNNs are dynamic in spatial resolution.
The question is why dynamic torch model is subject to original IR's static shape.

**Comment by cih9088** [comment]
> I'm not talking about IR model reshape, but torch model. (x) reshape IR to (w,h) -> convert to torch model (o) convert to torch model -> infer on (w,h) input
> 
> I understand that IR models have static shape. But torch CNNs are dynamic in spatial resolution. The question is why dynamic torch model is subject to original IR's static shape.

Converted torch models depend on IR graph. If IR graph has a hard-coded value related to input shape, that would apply to the torch model as well. 
To loosen shape check, we have to verify each IR has no hard-coded values.

File: mpa/modules/ov/graph/parsers/cls/cls_base_parser.py
Comment (wonjuleee) [inline mpa/modules/ov/graph/parsers/cls/cls_base_parser.py:53]: could you describe what False and 10 means?

File: mpa/modules/ov/graph/parsers/cls/cls_base_parser.py
Comment (wonjuleee) [inline mpa/modules/ov/graph/parsers/cls/cls_base_parser.py]: Isn't it good to move this to global variable at the top of the code?

File: mpa/modules/ov/graph/parsers/cls/cls_base_parser.py
Comment (wonjuleee) [inline mpa/modules/ov/graph/parsers/cls/cls_base_parser.py]: Isn't it good to move this to global variable at the top of the code?

File: mpa/modules/ov/graph/parsers/cls/cls_base_parser.py
Comment (wonjuleee) [reply mpa/modules/ov/graph/parsers/cls/cls_base_parser.py]: how about renaming this to `possible_neck_types`

File: mpa/modules/ov/graph/parsers/cls/cls_base_parser.py
Comment (wonjuleee) [inline mpa/modules/ov/graph/parsers/cls/cls_base_parser.py:115]: please add 
else:
    raise error

File: mpa/modules/ov/graph/parsers/cls/cls_base_parser.py
Comment (wonjuleee) [inline mpa/modules/ov/graph/parsers/cls/cls_base_parser.py:97]: Should "Result" be unique for the output layer name?
If not, how about set this to `possible_output_types` as a global variable?

File: mpa/modules/ov/graph/parsers/cls/cls_base_parser.py
Comment (cih9088) [reply mpa/modules/ov/graph/parsers/cls/cls_base_parser.py:115]: I made this function output `None` on purpose in case it fails to parse automatically because it is a utility function.
Error would be raised in [here](https://github.com/cih9088/model_preparation_algorithm/blob/feature/ov-otx/mpa/modules/ov/models/mmcls/backbones/mmov_backbone.py#L20-L26) where this function is actually used.

File: mpa/modules/ov/graph/parsers/cls/cls_base_parser.py
Comment (cih9088) [reply mpa/modules/ov/graph/parsers/cls/cls_base_parser.py:53]: `def bfs(self, node, reverse, depth_limit)`
The depth limit is set based on OMZ IR.

File: mpa/modules/ov/graph/parsers/cls/cls_base_parser.py
Comment (cih9088) [reply mpa/modules/ov/graph/parsers/cls/cls_base_parser.py:97]: `Result` is not a name but a node type and it is a part of head section.
This function tries to find neck component and output backbone, neck, head depending on input parameter.
I believe there would be no other types than `Result`.

File: mpa/modules/ov/graph/parsers/cls/cls_base_parser.py
Comment (cih9088) [reply mpa/modules/ov/graph/parsers/cls/cls_base_parser.py]: Fixed in https://github.com/openvinotoolkit/model_preparation_algorithm/pull/95/commits/e173066e7d7a43670452fb1b3dc9e99690c4503a

File: mpa/modules/ov/graph/parsers/cls/cls_base_parser.py
Comment (cih9088) [reply mpa/modules/ov/graph/parsers/cls/cls_base_parser.py]: Fixed in https://github.com/openvinotoolkit/model_preparation_algorithm/pull/95/commits/e173066e7d7a43670452fb1b3dc9e99690c4503a

**Comment by cih9088** [comment]
@harimkang FYI, you could pass `verify_shape=False` when initializing MMOV model.
For example,
```python
model_cfg = dict(
    type="ImageClassifier",
    backbone=dict(
        type="MMOVBackbone",
        model_path="omz://alexnet",
        remove_normalize=False,
        merge_bn=True,
        paired_bn=True,
        verify_shape=False,
        outputs="fc8/flatten_fc_input",
    ),
    neck=None,
    head=dict(
        type="MMOVClsHead",
        model_path="omz://alexnet",
        inputs="fc8/WithoutBiases",
        outputs="fc8",
        verify_shape=False,
        loss=dict(type="CrossEntropyLoss", loss_weight=1.0),
    ),
)
```

This would not verify that pytorch input shape does match to IR input shape

**Comment by harimkang** [comment]
> @harimkang FYI, you could pass `verify_shape=False` when initializing MMOV model. For example,

thank you, I'll check!

File: mpa/modules/ov/models/__init__.py
Comment (harimkang) [reply mpa/modules/ov/models/__init__.py:1]: image
I checked this again. I confirmed that unnecessary importing (& logs) do not come out by deleting the test import.
Could you please remove this? And how about moving this to the init of each folder (mmcls, mmdet, mmseg)?

Comment (cih9088) [reply mpa/modules/ov/models/__init__.py:1]: I can confirm that using public `mmseg` library, the warning messages is not appeared.
image
Do you want me to comment them out for the moment?

Comment (cih9088) [reply mpa/modules/ov/models/__init__.py:1]: Commented them out in  https://github.com/openvinotoolkit/model_preparation_algorithm/pull/95/commits/9d4dea34e7d1b146f339a5aefe9d46757238a18a

**Comment by harimkang** [comment]
@cih9088  could you fix the conflicting part?
@goodsong81 Do you agree to merge this PR after the conflict is resolved?

**Comment by cih9088** [comment]
@harimkang Rebased onto current `OTX`

**Review by goodsong81 — APPROVED** [review]

**Pull request merged**
